Source/SlimDetours/Thread.c

```diff
@@ -6,107 +6,133 @@
     _Out_ PULONG SuspendedHandleCount)
 {
     NTSTATUS Status;
-    ULONG i, ThreadCount, SuspendedCount;
-    PSYSTEM_PROCESS_INFORMATION pSPI, pCurrentSPI;
-    PSYSTEM_THREAD_INFORMATION pSTI;
-    PHANDLE Buffer;
-    HANDLE ThreadHandle, CurrentPID, CurrentTID;
-    OBJECT_ATTRIBUTES ObjectAttributes = RTL_CONSTANT_OBJECT_ATTRIBUTES(NULL, 0);
-
-    /* Get system process and thread information */
-    i = _1MB;
-_Try_alloc:
-    pSPI = (PSYSTEM_PROCESS_INFORMATION)detour_memory_alloc(i);
-    if (pSPI == NULL)
+    PHANDLE Buffer = s_Handles;
+    ULONG BufferCapacity = ARRAYSIZE(s_Handles);
+    ULONG SuspendedCount = 0;
+    BOOL CurrentThreadSkipped = FALSE;
+    HANDLE CurrentTID = (HANDLE)(ULONG_PTR)NtCurrentThreadId();
+    BOOL ClosePrevThread = FALSE;
+    HANDLE ThreadHandle = NULL;
+    while (TRUE)
     {
-        return STATUS_NO_MEMORY;
-    }
-    Status = NtQuerySystemInformation(SystemProcessInformation, pSPI, i, &i);
-    if (!NT_SUCCESS(Status))
-    {
-        detour_memory_free(pSPI);
-        if (Status == STATUS_INFO_LENGTH_MISMATCH)
+        HANDLE NextThreadHandle;
+        Status = NtGetNextThread(NtCurrentProcess(), ThreadHandle, THREAD_ACCESS, 0, 0, &NextThreadHandle);
+        if (ClosePrevThread)
         {
-            goto _Try_alloc;
+            NtClose(ThreadHandle);
         }
-        return Status;
-    }
 
-    /* Find current process and threads */
-    CurrentPID = (HANDLE)(ULONG_PTR)NtCurrentProcessId();
-    pCurrentSPI = pSPI;
-    while (pCurrentSPI->UniqueProcessId != CurrentPID)
-    {
-        if (pCurrentSPI->NextEntryOffset == 0)
+        if (!NT_SUCCESS(Status))
         {
-            Status = STATUS_NOT_FOUND;
-            goto _Exit;
+            if (Status == STATUS_NO_MORE_ENTRIES)
+            {
+                Status = STATUS_SUCCESS;
+            }
+            break;
         }
-        pCurrentSPI = (PSYSTEM_PROCESS_INFORMATION)Add2Ptr(pCurrentSPI, pCurrentSPI->NextEntryOffset);
-    }
-    pSTI = (PSYSTEM_THREAD_INFORMATION)Add2Ptr(pCurrentSPI, sizeof(*pCurrentSPI));
 
-    /* Skip if no other threads */
-    ThreadCount = pCurrentSPI->NumberOfThreads - 1;
-    if (ThreadCount == 0)
-    {
-        *SuspendedHandles = NULL;
-        *SuspendedHandleCount = 0;
-        Status = STATUS_SUCCESS;
-        goto _Exit;
-    }
+        ThreadHandle = NextThreadHandle;
+        ClosePrevThread = TRUE;
 
-    /* Create handle array */
-    Buffer = (PHANDLE)detour_memory_alloc(ThreadCount * sizeof(HANDLE));
-    if (Buffer == NULL)
-    {
-        Status = STATUS_NO_MEMORY;
-        goto _Exit;
-    }
+        if (!CurrentThreadSkipped)
+        {
+            THREAD_BASIC_INFORMATION BasicInformation;
+            if (!NT_SUCCESS(NtQueryInformationThread(
+                ThreadHandle,
+                ThreadBasicInformation,
+                &BasicInformation,
+                sizeof(BasicInformation),
+                NULL
+            )))
+            {
+                continue;

```

Comment:
`continue;` or `break;`, what do you think?

Reply:
Changed. I can't think of a reason for it to fail, but I tend to agree that it'd be better to know that something is wrong in this case.

On a related note, perhaps it's better to fail allocations if `RtlCreateHeap` fails instead of falling back to `NtGetProcessHeap` which may cause deadlocks. What do you think?

https://github.com/KNSoft/KNSoft.SlimDetours/blob/4cc236ae98ef10788da7a2b08aaf3bf660281bca/Source/SlimDetours/Memory.c#L96-L101

Reply:
I prefer `break` because the current thread is not skipped yet, to avoid suspending self is more worthwhile. `NtQueryInformationThread` fails and a unsuspended thread executes invalid instruction while we modifying instructions is the lowest risk. There seems no 100% safe, we strive for it as much as we can.

Reply:
Unless I misunderstood what you meant:
It couldn't suspend itself before, too. The difference is that before the change (`continue`), threads which couldn't be queried would be skipped, which could cause a race condition. With the latest change (`break`), the operation fails in this case, so starting the transaction will fail.

Reply:
Sorry, I was careless when I read the code, I'll revert it to `continue` in next commit.

Reply:
Reverted to `continue;` in https://github.com/KNSoft/KNSoft.SlimDetours/commit/b2f9fbaba736973acb0b28a2fa4c8c3a4169b229

As for the second question, I think a very small risk is better than not available. I prefer to add `SlimDetoursInitialize` to let user know the result of initialization process, returns a warning (rather than error) if the fall back occurred, then user could choose give up hooking or take on small risk. This means change current interfaces, add and `SlimDetoursUninitialize` (you suggested in #8 ).